src/simd_basic.h

```diff
@@ -0,0 +1,404 @@
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef __QUAD_SIMD_BASIC_H__
+#define __QUAD_SIMD_BASIC_H__
+
+#include <x86intrin.h>
+
+namespace quadiron {
+namespace simd {
+
+/* ================= Basic Operations for u32 ================= */
+
+/**
+ * Modular addition for packed unsigned 32-bit integers
+ *
+ * @param x input register
+ * @param y input register
+ * @param q modulo
+ * @return (x + y) mod q
+ */
+inline VecType ADD_MOD(VecType x, VecType y, uint32_t q)
+{
+    VecType res = ADD32(x, y);
+    return MINU32(res, SUB32(res, CARD(q)));
+}
+
+/**
+ * Modular subtraction for packed unsigned 32-bit integers
+ *
+ * @param x input register
+ * @param y input register
+ * @param q modulo
+ * @return (x - y) mod q
+ */
+inline VecType SUB_MOD(VecType x, VecType y, uint32_t q)
+{
+    VecType res = SUB32(x, y);
+    return MINU32(res, ADD32(res, CARD(q)));
+}
+
+/**
+ * Modular negation for packed unsigned 32-bit integers
+ *
+ * @param x input register
+ * @param q modulo
+ * @return (-x) mod q
+ */
+inline VecType NEG_MOD(VecType x, uint32_t q)
+{
+    VecType res = SUB32(CARD(q), x);
+    return MINU32(res, SUB32(res, CARD(q)));
+}
+
+/**
+ * Modular multiplication for packed unsigned 32-bit integers
+ *
+ * @note We assume that at least `x` or `y` is less than `q-1` so it's
+ * not necessary to verify overflow on multiplying elements
+ *
+ * @param x input register
+ * @param y input register
+ * @param q modulo
+ * @return (x * y) mod q
+ */
+inline VecType MUL_MOD(VecType x, VecType y, uint32_t q)
+{
+    VecType res = MUL32(x, y);
+    VecType lo =
+        (q == F3) ? BLEND8(ZERO, res, MASK8_LO) : BLEND16(ZERO, res, 0x55);
+    VecType hi = (q == F3) ? BLEND8(ZERO, SHIFTR_1(res), MASK8_LO)
+                           : BLEND16(ZERO, SHIFTR_2(res), 0x55);
+    return SUB_MOD(lo, hi, q);
+}
+
+/**
+ * Modular general multiplication for packed unsigned 32-bit integers
+ *
+ * @note It's necessary to verify overflow on multiplying elements
+ *
+ * @param x input register
+ * @param y input register
+ * @param q modulo
+ * @return (x * y) mod q
+ */
+inline VecType MULFULL_MOD(VecType x, VecType y, uint32_t q)
```

Comment:
This is micro-optimization as it's crucial for performance. `MUL_MOD` is used in FNT, while `MULFULL_MOD` in NF4

Reply:
I test for FNT with your proposition, i.e. using optional argument `filter`: encoding speed ~2050MB/s
My current implementation without `filter` gives about ~2150MB/s
I have tested the template parameter way yet.

Reply:
So almost 5%, quite costly…
A drop in perf was to be expected because of the branch that may be hard to predict, the template version should perform better.

Reply:
The new code is more readable and with less duplication, good job!